### ISSUE TITLE
rclone not working with ssh-agent, possibly linked to macOS 10.12

### ISSUE THREAD

**Issue by kubark42** [ncw/rclone#1218]
> What is your rclone version (eg output from `rclone -V`)

rclone-v1.36

> Which OS you are using and how many bits (eg Windows 7, 64 bit)

macOS 10.12.3

> Which cloud storage system are you using? (eg Google Drive)

SFTP

> The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

Any `rclone` command depending on `ssh-agent`.

> A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

```
rclone-v1.35-158-g4652db3β-osx-amd64/rclone lsd seit:
2017/03/06 14:33:52 Failed to create file system for "seit:": couldn't connect ssh: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
```

And yet `ssh remote_server_foo.bar.com` works perfectly with the key, i.e. no password request.

Here's the config:
```
[seit]
type = sftp
host = remote_server_foo.bar.com
user = kenz
port = 
pass = 
--------------------
Edit remote
Value "host" = "remote_server_foo.bar.com"
Edit? (y/n)>
y) Yes
n) No
```

After it didn't work, I edited the config to add a password and it worked again. So I'm fairly sure I configured everything else properly.

Links to #521.

**Comment by ncw** [comment]
What happens if you do

```
$ env | grep ssh
```

Do you see something like this?

```
SSH_AUTH_SOCK=/tmp/ssh-Sj7JaoojWnEb/agent.3807
```

Do you see your keys if you do `ssh-add -l`?

I wonder If I'm using an outdated way of getting the agent [see blog post](http://blog.ralch.com/tutorial/golang-ssh-connection/)

**Comment by kubark42** [comment]
```
$ env | grep -i ssh
SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.3wxD7DaKDY/Listeners
```

```
$ ssh-add -l
The agent has no identities.
```

Oooohhh... Time to read that blog entry!

```
$ eval "$(ssh-agent -s)"
Agent pid 54543
```

```
$ ssh-add -l
2048 SHA256:foo/BAR2342343fasdf ~/.ssh/id_rsa (RSA)
```

Hey, this time something's there. Maybe it'll work?

```
$ rclone-v1.35-158-g4652db3β-osx-amd64/rclone lsd seit:
          -1 2014-02-24 16:57:13        -1 .cache
          -1 2014-05-20 20:41:00        -1 my_project
          -1 2017-03-06 14:28:57        -1 .ssh
          -1 2014-05-20 20:42:36        -1 my_project.git
```
Success!

It seems like `ssh-agent` is an extra step which users will have to (know to) run. I don't know if `rclone` has any control over this, but would it be possible for it to launch `ssh-agent` if it isn't already running?

**Comment by ncw** [comment]
Thanks for working that out.  There must be a special macOS way of doing this...

A bit of searching [brings up this thread](https://github.com/lionheart/openradar-mirror/issues/15361) which has several work-arounds in it, probably the best is described [in this blog post](http://joshbuchea.com/revert-ssh-agent-behavior-pre-macos-sierra/)

**Comment by howardm** [comment]
Hi,

I, too, report this problematic issue with v 1.36

I've configured a SFTP remote to my remote Mac mini, specifying (passwordless) SSH.

I'm trying to connect to my remote Mac mini (macOS Sierra 10.12.1) from my laptop (macOS Sierra 10.12.3).

It fails with this message:

howardm$ rclone lsd Macmini:
2017/03/19 20:50:59 Failed to create file system for "Macmini:": couldn't connect ssh: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain

I can, obviously, connect to it otherwise via SSH in the usual way.

I'm afraid I cannot understand the "workaround" described above.

Howard

**Comment by kubark42** [comment]
@howardm there's definitely a bit of an issue with the workflow difference between doing `ssh` with keys and doing `ssh-agent`. I have found that the steps which work for me with Matlab are to

```
# It's important that you use `eval()`. Silly, I know, but that's the way ssh-agent works.
eval "$(ssh-agent -s)"

# Point toward your private ssh key. Mine was in ~/.ssh/id_rsa
ssh-add -K ~/.ssh/id_rsa 
.
.
.
# stuff here, like `rclone lsd Macmini:`
.
.
.

# Again, important to use `eval()`
eval "$(ssh-agent -k)"
```

**Comment by howardm** [comment]
Hmm. I'm very much a newbie at this, and still don't understand the process.

When I generated my private/public key combination, I didn't specify a passphrase (as best I remember). I certainly don't have to enter a passphrase for all the other `rsyncs` I use my pair for.

Thus, I don't understand the need for a ssh-agent here. Or, is this process needed here for rClone for some reason ?

___

As in: 

![01](https://cloud.githubusercontent.com/assets/6101226/24179956/dda40fbe-0e77-11e7-8060-053fa5f34763.png)

[From: https://help.github.com/articles/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent/]

Also, if I add it to ssh-agent (as described in the `Tip` in the screenshot above), is that a one-time thing/requirement only ?

Cheers,

Howard

**Comment by kubark42** [comment]
> Thus, I don't understand the need for a ssh-agent here. Or, is this process needed here for rClone for some reason ?

`rclone` uses it, but I have limited experience with ssh so will defer to @ncw for explanations.

> Also, if I add it to ssh-agent (as described in the Tip in the screenshot above), is that a one-time thing/requirement only ?

You'll have to run the code I pasted every time you open a new terminal. It can get annoying because if you close the terminal without killing ssh-agent, it will continue to run in the background and it will never end. I solve that problem by periodically using `killall ssh-agent`.

**Comment by ncw** [comment]
Would it be helpful if rclone could use a key file directly without using ssh-agent?

**Comment by kubark42** [comment]
@ncw I *think* so, but I might be misunderstanding what that means. The ideal situation would be like with Google Drive, where you set things up at config time and then everything is self contained.  But I'd be fine with it taking my id:password on the command line, this would probably be the least amount of effort since this can be scriptable. Not sure if that's a realistic option.

How do tools like `ssh`, `git`, and `rsync` work? Do they load ssh-agent themselves, and make a best guess as to where the SSH keys are?

And if the same functionality can be gotten, could `rclone` give help to the user, maybe check the connection and inform the user how to make/transfer keys if the SSH connection fails?

**Comment by ncw** [comment]
I meant as a config option you would give a path to a key file, and rclone would load that instead of using the ssh-agent.  rclone could even copy the key file into the config file.

> How do tools like ssh, git, and rsync work? Do they load ssh-agent themselves, and make a best guess as to where the SSH keys are?

They would expect an ssh-agent to be running

**Comment by kubark42** [comment]
> I meant as a config option you would give a path to a key file, and rclone would load that instead of using the ssh-agent. rclone could even copy the key file into the config file.

That would definitely work. Copying the key into the config file would be a cool, because then when I'm deploying my uploader tool I could copy only the conf file, instead of both it and the ssh key.

> They would expect an ssh-agent to be running

Gotcha. So I did a simple test to try to better understand how this works. When I `kill-all ssh-agent` and run `ssh foo.bar`, it loads ssh-agent. And when I open up another terminal, it uses the same ssh-agent. `git` does the same. So could `rclone` also launch `ssh-agent` and get keys in the same manner?

**Comment by kubark42** [comment]
> So could rclone also launch ssh-agent and get keys in the same manner?

Oooh, interesting: http://rabexc.org/posts/pitfalls-of-ssh-agents. **Too many keys, github, and friends**.

So it looks like the most robust approach might be to have `rclone` load a new ssh-agent, add the key (perhaps saved in the config), and then kill it on exit.

**Comment by howardm** [comment]
FYI

I've used this method successfully:

1. ssh-add -K path/to/my_key  [needed once only]

2. Use this bash script for my scheduled backups

___
![pic](https://cloud.githubusercontent.com/assets/6101226/24276365/feed4b2e-0ffb-11e7-9722-dad344b6c290.jpg)
___

So the agent is killed after the sync or copy has finished.

Howard

**Comment by ncw** [comment]
@kubark42 
> So I did a simple test to try to better understand how this works. When I kill-all ssh-agent and run ssh foo.bar, it loads ssh-agent. And when I open up another terminal, it uses the same ssh-agent. git does the same. So could rclone also launch ssh-agent and get keys in the same manner?

I wonder where that behaviour of launching ssh-agent comes from because it certainly isn't standard ssh behaviour on linux.  I launch my ssh-agent once and I have to type in a very long password - it would be might annoying to have to do that on every rclone invocation!

I think this must be something to do with apple's integration of ssh keys into the secure keychain - that is what the `ssh-add -K` does in @howardm 's writeup above.  So I guess ssh-agent is just started as a shim for the secure keychain.  This must be OS X specific behaviour...

My theory is that this is done by the standard library.  When I cross compile rclone (which I do for the releases) it doesn't use the stdlib so you don't get this behaviour.  However if you compile rclone from source directly on OS X you will get this behaviour.

Can anyone confirm that?  If you need help compiling rclone from source there are some [brief instructions in the docs](http://rclone.org/install/#install-from-source).

If that is the case then I can look at getting the travis build to build the OSX binary...

**Comment by kubark42** [comment]
> Can anyone confirm that? If you need help compiling rclone from source there are some brief instructions in the docs.

I just now built it from source and while it did launch `ssh-agent`, it was not able to connect. Adding the keys as per the above discussion, however, did allow the built-from-source binary to connect.

It seems like SSH key handling is annoyingly OS specific.

**Comment by howardm** [comment]
I don't know if this will provide some helpful diagnostic information:

I followed the [instructions in this post.](http://joshbuchea.com/revert-ssh-agent-behavior-pre-macos-sierra/) about the ssh config file.

I thought I would then *not* have to use `ssh-add -A` in my script, as I mention above.

But my script doesn't work if it's not there.

Howard

**Comment by kubark42** [comment]
@ncw, I wasn't able to make this work in Windows. I get a `socket: An address incompatible with the requested protocol was used.` error. This is similar to https://github.com/hashicorp/terraform/issues/3423, with specific details at https://github.com/hashicorp/terraform/issues/3423#issuecomment-153447080.

https://github.com/hashicorp/terraform/pull/4323/files was a very easy fix for Terraform, and in Go to boot! But not sure if it's relevant here.

**Comment by ncw** [comment]
@kubark42 thanks for trying the rclone built from source. So after you've done `ssh-add -K` it then works correctly, starting and stopping an ssh-agent - is that right?  That sounds like it is working properly.

re Windows: interesting!  Can you make a separate issue with that please.  And from a brief look a the code, I think the terraform fix will work for rclone too.

**Comment by kubark42** [comment]
> So after you've done ssh-add -K it then works correctly, starting and stopping an ssh-agent - is that right? That sounds like it is working properly.

Yes, that's it exactly. However, other macOS tools which automatically launch `ssh-agent` also load the ssh keys, so this behavior diverges slightly from that. 

> re Windows: interesting! Can you make a separate issue with that please. And from a brief look a the code, I think the terraform fix will work for rclone too.

Yes, will do.

**Comment by diamondsw** [comment]
Jumping into an issue near and dear to my heart (i.e., banging my head against it ;) ). macOS Sierra apparently made the decision to bring their SSH implementation in-line with upstream OpenSSH 7.3. This resulted in many things breaking related to SSH, including the way we'd been using ssh-agent, Unicode support between macOS systems (by default, it no longer passes the LC_ environment, breaking non-ASCII support), etc. On macOS I got around it similarly to the above, wherein every SSH session I have to start with:
`eval `ssh-agent -s` && ssh-add -A`

That starts up a new SSH agent, adds the requisite environment variables, and adds my standard identity file (~/.ssh/id_rsa). However, if I forget to do this then every SFTP rclone fails. I haven't even tried to get this running via launchd as I'm positive ssh-agent will fail.

Then I started trying to get this working on my new Synology (now that I have a NAS capable of running backups directly from). It doesn't even have ssh-agent, and no signs it will be added. So even though I have my keys all ready and available., I can't use them because of the lack of an agent.

> Would it be helpful if rclone could use a key file directly without using ssh-agent?

It would certainly solve both of my problems, and I can't really see a downside. People can use ssh-agent it it's working for them, but if it's not then fall back on specifying the key. Definitely better than hardcoding a password in the rclone config file (although that could be a fallback of last resort), and some systems (/me raises hand) disallow non-public key auth due to constant floods of password hacking attempts (if my logs could tell tales... they'd probably all be in Chinese, based on the IP ranges).

**Comment by ncw** [comment]
I've updated [the docs](https://github.com/ncw/rclone/blob/master/docs/content/sftp.md#ssh-agent-on-macos) with a combination of @diamondsw  and @howardm 's suggestions.

I've implemented ssh key support in #1494 

So I'm not sure what else I can do - any ideas?

**Comment by ncw** [comment]
I'm going to close this now as the docs are updated and we now have ssh key support.
